sys/pool.c

```diff
@@ -0,0 +1,203 @@
+#include <stdc.h>
+#include <vm.h>
+#include <physmem.h>
+#include <queue.h>
+#include <bitstring.h>
+#include <common.h>
+#include <pool.h>
+
+#define PI_MAGIC_WORD 0xcafebabe
+#define PP_FREED_WORD 0xdeadbeef
+
+typedef struct pool_slab {
+  LIST_ENTRY(pool_slab) ph_slablist; /* pool slab list */
+  vm_page_t *ph_page;                /* page containing this slab */
+  uint16_t ph_nused;                 /* # of items in use */
+  uint16_t ph_nfree;  /* # of free (and available) items, a bit redundant but
+                         there would be padding anyway */
+  uint16_t ph_ntotal; /* total number of chunks*/
+  uint16_t ph_start;  /* start offset in page */
+  bitstr_t ph_bitmap[0];
+} pool_slab_t;
+
+typedef struct pool_item {
+  uint32_t pi_canary;   /* PI_MAGIC_WORD by default, if modified then memory
+                           corruption has occured */
+  pool_slab_t *pi_slab; /* pointer to slab containing this item */
+  unsigned long pi_data[0];
+} pool_item_t;
+
+static pool_item_t *get_pi_at_idx(pool_slab_t *slab, unsigned i, size_t size) {
+  return (pool_item_t *)((intptr_t)slab + slab->ph_start +
+                         i * (size + sizeof(pool_item_t)));
+}
+
+static pool_slab_t *create_slab(pool_t *pool) {
+  /* log("create_slab: pool = %p, pp_itemsize = %d", pool, pool->pp_itemsize);
```

Comment:
Instead of commenting it out I suggest to use `debug` macro:
```c
#ifdef DEBUG
#define debug(...) log(__VA_ARGS__)
#else
#define debug(...)
#endif